segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/stat/AutoScaleProcessor.java

```diff
@@ -61,10 +66,12 @@
 
     private final EventStreamClientFactory clientFactory;
     private final Cache<String, Pair<Long, Long>> cache;
-    private final AtomicReference<EventStreamWriter<AutoScaleEvent>> writer;
+    @GuardedBy("$lock")
+    private CompletableFuture<EventStreamWriter<AutoScaleEvent>> writer;
```

Comment:
Non-final fields make it look like something's not right. Let's see if we can make this better. Would this work?
1. Make this final and remove the "GuardedBy"
2. In the `getWriter` method, do exactly what you're doing, but when the retry loop is done, set the result on this future, whether it be the writer or an exception. You can use `Futures.completeAfter` (I may forget the actual name). 

This way we don't need a new lock and we can make this field final. `CompletableFuture` internally is thread safe so it takes care of any concurrency issues for us.

Reply:
We can avoid non final writer. But the objective is to make the bootstrap of writer be triggered exactly once which will eventually complete the future. 
In existing code, the bootstrap was being invoked from the constructor.
But now we want to do it lazily, upon first request. But there could be multiple concurrent requests. 
So we need some syncrhonization to make sure exactly one of them end up attempting to create a writer. 
So syncrhonization is really for that. And instead of using a non-final boolean which is synchronized (or atomic boolean) i had chosen to do with writerFuture = null. 

I have an alternate idea though -- which would still require using atomic boolean.. let me make and push that change and see if that is better.